docs/spec/reactors/consensus/proposer-selection.md

```diff
@@ -2,45 +2,268 @@
+When a new validator is added same problem as the one described for removal appears, the sum of priorities in the new set is not zero. This is now fixed with the shift step introduced above.
+
+One other issue that needs to be addressed is the following. A validator V with low voting power that has just been elected is moved to the end of the queue. If the validator set is large and/ or with other validators with significantly higher power, validator V will have to wait a lot of rounds to be elected. If V could remove and re-add itself to the validator set, it would make a significant (albeit unfair) "jump" ahead in the queue, decreasing the numbers of rounds it has to wait. 
+
+In order to prevent this, when a new validator is added, its initial priority is not set to 0 but to a lower value. This value is determined from the simulation of the above scenario: it is assumed that V had just proposed a block and was at the back of the queue when it was removed and re-added. In addition, to discourage intentional unbound/ bound operations, a penalty factor is applied.
+
+If `vset = {v1,..,vn, V}` and `V` would be selected as proposer then it would have gone through the following changes:
+
+    ...
+      A(V)+=VP(V) // part of first for loop
+    ...
+    A(V)-=P  
+
+It follows that `A(V)-=sum(VP(i) for i in {v1,..,vn})`
+
+Therefore when `V` is added to the `{v1,..,vn}` set its initial priority will be:
+
+    A(V) = -sum(VP(i) for i in {v1,..,vn})
+
+Curent implementation uses a penalty factor of 1.125, so the initial priority for a newly added verifier is:
+    
+    A(V) = -1.125 * P
+
+If we consider the validator set where p3 has just been added:
+
+Validator | p1 | p2 | p3
+----------|--- |--- |---
+VP        | 1  | 3  | 8
+
+Let's assume that the last round was R and the proposer priorities were as shown in first row (R end) with their sum being 0. p3 is added with A(p3) = -4 (penalty loss due to integer division?)
+In the next round, p3 will still be ahead in the queue, elected as proposer and move back in the queue (same position it was just added)
+
+|Round Priority  | -4 | -3 | -2 | -1 | 0  | 1 | 2 | 3 | 4 | Alg step
+|--------------- | ---|--- |----|--- |--- |---|---|---|---|--------
+|H/R end         |    |    | p2 |    |    |   | p1|   |   |add p3
+|                | p3 |    | p2 |    |    |   | p1|   |   |A(p3) = -4
+|H+1/0           |    |    |    |    |    | p2|   | p1| p3|A(i)+=VP(i)
+|                | p3 |    |    |    |    | p2|   | p1|   |A(p3)-=P
+
+### Proposer Priority Range
+With the introduction of shifting, an interesting case occurs. Low power validators that bind early benefit from subsequent addition of high power validators. This is because new validators are added with negative priorities and this causes an increase in priority for all other validators during the shifting operation. As a consequence, low power validators may end up with high priority and election algorithm will fail with respect to fairness.
+
```

Comment:
This case does not invalidate fairness as fairness is defined only as an eventual property (during period of stability). Adding or removing validators will not break the fairness if validator set stabilise; the expected sequence (that corresponds to the voting power of each validator) will just appear after some number of rounds.

Reply:
I think the justification for the averaging is to keep the prios somewhat closer to zero. 

The justification for normalization is that without it priorities can grow (unironically due to averaging) into both directions. Validators that join early after a large power validator will accumulate lots of priority (relative to validators that join later). This penalizes freshly joined validators more than necessary. With each increment step a proposer get de-priotized in `incrementProposerPriority` by:
```go
mostest.ProposerPriority = safeSubClip(mostest.ProposerPriority, vals.TotalVotingPower())
``` 
But as the very large validator disappeared subtracting total voting power isn't much compared to the gained prio (from the existence of the early large combined with the averaging). 
Sure, this effect *eventually* wears off. But it potentially only wears off after many, many rounds where the early validator(s) were proposer (even with no power changes in the mean time). 

OK, this is not a formal requirement but I guess it can be seen as unfair if early (probably smaller) validators get to propose much more often until (eventually) this becomes fair again.

Reply:
I think @cwgoes came up with this property:

> For any sequence of proposer elections with no power changes and n blocks, the difference between n*power/totalPower and the number of proposed blocks (for each validator) is at most 1.

I think this property makes sense independent from out implementation. It basically says that a validator doesn't get to propose more than it should for any sequence of elections with no power changes (and some number of blocks n).

FWIW this property would break if we only do averaging and no normalization (the difference would be larger for some sequence of elections with no power changes as described in #2960).

Does this make any sense?

Reply:
I think @cwgoes' requirement is R2 exactly (for C=1). My question was how to formulate the same requirement for the proposer selections with power changes.
Just to trigger some more thoughts, I added this C=1/2 just because we double the min/max distance to 2 * voting power. Why was this 2 chosen? And why don't limit to just voting power?